multi: readonly and admin macaroon session

ellemouton opened the pull request:
- part of https://github.com/lightninglabs/lightning-node-connect/issues/17
- depends on https://github.com/lightninglabs/lightning-node-connect/pull/27

With this PR, we can create a admin or readonly macaroon session type. This bakes a macaroon for a session and adds it to the Noise `authData` payload so the client can use the mac to authorise requests instead of just the UI password.

After this is in, the UI can start defaulting to creating a readonly session and can add an option for the user to select that they want to create an admin session. Once that is done (ie, there is no way to create a UIPassword session anymore), we can add a DB migration diff that downgrades all existing UIPassword sessions to Readonly sessions

ellemouton commented on session/tlv.go:
```diff
@@ -150,5 +147,46 @@ func DeserializeSession(r io.Reader) (*Session, error) {
+
 	return session, nil
 }
+
+func macaroonPermsToString(perms []bakery.Op) string {
+	var macRecipe string
+	for _, op := range perms {
+		if macRecipe != "" {
+			macRecipe += ";"
+		}
+
+		macRecipe += fmt.Sprintf("%s:%s", op.Entity, op.Action)
+	}
+
+	return macRecipe
+}
+
+func macaroonPermsFromString(macRecipe string) ([]bakery.Op, error) {
+	perms := strings.Split(macRecipe, ";")
+	recipe := make([]bakery.Op, 0, len(perms))
+	for _, permStr := range perms {
+		perm := strings.Split(permStr, ":")
+		if len(perm) != 2 {
+			return nil, errors.New("invalid macaroon " +
+				"recipe encoding")
+		}
+
+		recipe = append(recipe, bakery.Op{
+			Entity: perm[0],
+			Action: perm[1],
```
gonna replace all this. just found tlv.MakeDynamicRecord. gonna use that instead. also need to add caveates

ellemouton commented:
@guggero , i have updated this to include the readonly session type. I think the DB migration diff should be in a separate PR after the UI changes that allow you to pick a session type have been done.

guggero commented on session/tlv.go:
```diff
@@ -152,3 +154,178 @@ func DeserializeSession(r io.Reader) (*Session, error) {
+				return err
+			}
+
+			w.Write(permBytes)
```
nit: unchecked return value here and a few places below.

guggero commented on session/tlv.go:
```diff
@@ -152,3 +154,178 @@ func DeserializeSession(r io.Reader) (*Session, error) {
+		// we then convert to a byte slice. We first write the length
+		// of the byte slice, followed by the byte slice.
+		for _, perm := range v.Permissions {
+			permBytes := []byte(permissionToString(perm))
```
Nice simplification! I guess we could even have concatenated all permission pairs together, separated by a comma. But this is also good and adds one less restriction on the name/content of the entity/action pair.

guggero commented on session/tlv.go:
```diff
@@ -152,3 +154,178 @@ func DeserializeSession(r io.Reader) (*Session, error) {
+			// Write the length of the caveat verification ID
+			// followed by the verification ID itself.
+			err = tlv.WriteVarInt(
+				w, uint64(len(caveat.VerificationId)), buf,
```
Ah I should have mentioned that we're not really going to add third-party caveats any time soon (probably). So the `VerificationId` and `Location` aren't really needed. But let's leave them in now that you've added them, so we're future proof.

guggero commented on session_rpcserver.go:
```diff
@@ -82,18 +89,36 @@ func (s *sessionRpcServer) resumeSession(sess *session.Session) error {
+		if err != nil {
+			log.Debugf("Not resuming session %x. Could not bake"+
+				"the necessary macaroon: %w",
+				sess.LocalPublicKey.SerializeCompressed(), err)
```
return `nil` after logging.

guggero commented on session_rpcserver.go:
```diff
@@ -20,33 +21,39 @@ type sessionRpcServer struct {
 	}
 
+	var macPerms []bakery.Op
+	if typ == session.TypeMacaroonAdmin {
```
Unify this `if` over the `typ` with the one above and transform into a nice `switch` statement that combines both tasks?

guggero commented on session_rpcserver.go:
```diff
@@ -20,33 +21,39 @@ type sessionRpcServer struct {
 
+	var macPerms []bakery.Op
+	if typ == session.TypeMacaroonAdmin {
+		macPerms = getAllPermissions()
```
I'm still trying to decide if we should define these permissions at _session create time_ or at _bake time_. E.g. if we add a new permission to one of the daemons, does the user expect the session to upgrade automatically (since it's an admin or full read-only session) after restart? Or would the user need to create a new session if something in the permissions changed? The first would be better UX while the second might be a bit more security conscious.

The difference in the code would be: for `TypeMacaroonAdmin` and `TypeMacaroonReadonly` we wouldn't store the recipe in the DB (that would be reserved for the future `TypeMacaroonCustom`) and assemble the list of permissions right before baking the macaroon.

What do you think?

ellemouton replied:
hmmmm yeah I agree with your pro's and con's list of the two options. I guess option 1 is ok given that if the user wants to be more security focused and not let the perms list auto-update, they should use the custom option

ellemouton replied:
option 1 also makes the db migration easier since all we then need to do is change the type & dont need to go populate the recipe field

guggero reviewed (commented):
Nice, one step closer to getting rid of the UI password :tada:
Did a quick pass and the code looks good! Will do some manual testing on the second pass. In the meantime, we should probably discuss the question of when to assemble the permission list (see inline comment).

ellemouton commented:
Thanks for the review! I have updated accordingly and went for "at bake time" approach.

guggero reviewed (commented):
Code looks very good, great feature update!
I tested it out locally and found that the newly baked super macaroons don't work for faraday/loop/pool calls since they're not recognized as super macaroons.
I'll send you a patch offline that you can add to the PR to fix that.

guggero commented:
Ah, we also discussed adding a DB migration to the session DB to migrate over any existing UI password sessions to admin macaroon ones. But I don't think we can do that just yet, since first we need a way to make the super macaroon work with remote faraday/loop/pool. I have an idea of how to do that, can discuss it offline.

lightninglabs-deploy commented:
@roasbeef: review reminder

lightninglabs-deploy commented:
@roasbeef: review reminder

Roasbeef commented on session/interface.go:
```diff
@@ -39,7 +47,7 @@ type Session struct {
 	ServerAddr      string
 	DevServer       bool
 	MacaroonRootKey uint64
-	Macaroon        *macaroon.Macaroon
```
Do we need to do a "migratIon" to properly make this change? Or seems TLV saves the day here...

Roasbeef commented on session/tlv.go:
```diff
@@ -58,17 +67,6 @@ func SerializeSession(w io.Writer, session *Session) error {
 		),
 	}
 
-	if session.Macaroon != nil {
```
Re backwards compat, would appear we always need to continue reading this. 

With the way TLV works, eventually the new clients just stop writing it and the logic checks to see which of the two have been populated.

Roasbeef commented on session/tlv.go:
```diff
@@ -109,12 +115,15 @@ func DeserializeSession(r io.Reader) (*Session, error) {
 		tlv.MakePrimitiveRecord(
 			typeMacaroonRootKey, &session.MacaroonRootKey,
 		),
-		tlv.MakePrimitiveRecord(typeMacaroon, &macaroonBytes),
```
See comment re still reading this.

Roasbeef commented on session/tlv.go:
```diff
@@ -131,14 +140,7 @@ func DeserializeSession(r io.Reader) (*Session, error) {
 	session.ServerAddr = string(serverAddr)
 	session.DevServer = devServer == 1
-
-	if t, ok := parsedTypes[typeMacaroon]; ok && t == nil {
```
Here too

Roasbeef commented on session/tlv.go:
```diff
@@ -152,3 +154,186 @@ func DeserializeSession(r io.Reader) (*Session, error) {
 }
+
+// encodeMacaroonRecipe serializes a macaroon recipe so that it can be stored
+// as a tlv value. The serialization will follow the following pattern:
```
If we want to make this encoding _more_ flexible. We can us a _nested_ TLV instance here. Sounds scary, but it just means adding a length prefix before the inner TLV blob, and doing TLV as normal within that.

ellemouton replied:
ok nice! i decided to go all out on the nested-tlv thing after seeing this comment. Let me know if it is too much now 😅

Roasbeef commented on terminal.go:
```diff
@@ -141,7 +141,8 @@ type LightningTerminal struct {
 
-	lndClient *lndclient.GrpcLndServices
+	lndClient   *lndclient.GrpcLndServices
+	basicClient lnrpc.LightningClient
```
Can likely start to retroactively move some of the new macaroon methods in to a new sub-server, but that's outside the scope of this PR.

Roasbeef commented on subserver_permissions.go:
```diff
@@ -55,13 +55,28 @@ func getAllMethodPermissions() map[string][]bakery.Op {
 
+// denyList is the list of permission actions that will be denied for readonly
+// macaroons.
+// TODO(elle): the individual daemon libraries should export a readonly
```
Which can be done mainly by exporting the perms in the various `*rpc` packages?

Roasbeef replied:
Alternatively, this can function as a white list: iirc they all have "read" in there as the perms right?

ellemouton replied:
> Which can be done mainly by exporting the perms in the various *rpc packages?

yes

> Alternatively, this can function as a white list: iirc they all have "read" in there as the perms right?

good idea 👍

guggero replied:
I think this might cause an issue with Loop since some calls require a `loop:in` or `loop:out` permission in addition to the `terms:read` one (for example `LoopOutTerms`). But we can adjust that later if we find that we need loop terms in read-only mode.

Roasbeef commented on cmd/litcli/sessions.go:
```diff
@@ -50,6 +50,14 @@ var addSessionCommand = cli.Command{
+				"determine the permissions a user has when " +
+				"connecting with the session. Options " +
+				"include readonly|admin",
+			Value: "readonly",
```
We'll need to update our docs here as the UI assumes it has all perms (?) atm. 

I think a good slot to settle in would be normal perms w/ certain thresholds.

Roasbeef commented on session/macaroon.go:
```diff
@@ -0,0 +1,76 @@
+const (
+	// SuperMacaroonRootKeyBitmask is the prefix bitmask we set on a super
+	// macaroon's root key to clearly mark it as such.
+	SuperMacaroonRootKeyBitmask uint64 = 0xFFEEDDCC_00000000
```
Where's this come from? 

Couldn't we alternatively go with a prefix? Or maybe I don't understand the implications of the ID: I thought it was just a top-level key we used to read the macaroon secret to validate incoming macaroons?

guggero replied:
We need a way to identify a super macaroon by just reading the macaroon itself. Since a super macaroon is always baked by lnd, we can specify the _ID_ of the root key to use. And that ID (not the root key itself, obviously) is encoded within the macaroon (in a field called `storage_id`). Since the ID is `uint64`, we have 8 bytes to work with and this mask is effectively the prefix we use for any super macaroon.
I guess we could've simplified the code somewhat by generating 4 random bytes, prepending the prefix and only afterwards convert it to `uint64` instead of doing bit mask magic...

Roasbeef reviewed (changes requested):
Completed an initial pass, main thing I still need to grok is the implications re the identification of the super macaroons. 

I think we should land https://github.com/lightninglabs/lightning-terminal/pull/294 before this, so we ensure we're able to add adequate end to end test coverage.

lightninglabs-deploy commented:
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@ellemouton, remember to re-request review from reviewers when ready

ellemouton commented:
!lightninglabs-deploy mute

ellemouton commented:
muting this until the dependent PRs are complete. (should probably add an option to lbot so that it can depend on prs outside of the current repo)

ellemouton reviewed (commented):
Thanks for the review @Roasbeef 🚀 
See my comments on not reading the macaroon tlv anymore. tl;dr: we never wrote anything there before, so i think we dont need to keep reading from it to remail backwards compatible

ellemouton commented:
I have pushed a new version based on the latest [dynamic auth pr](https://github.com/lightninglabs/lightning-node-connect/pull/27) & addressed some of the comments in the previous review. Note that for testing: LNC on the client must be updated to the dynamic auth version

guggero reviewed (commented):
A couple of comments about the TLV encoding. Other than those, the PR looks good to me!

And I also simplified my commit (cc34d6e8) a bit according to @Roasbeef's comments. See the attached diff. Thanks for applying it to this PR!
[rootkey.diff.txt](https://github.com/lightninglabs/lightning-terminal/files/7995139/rootkey.diff.txt)

ellemouton commented on session/tlv.go:
```diff
@@ -152,3 +167,336 @@ func DeserializeSession(r io.Reader) (*Session, error) {
+
+			// We encode the record with a varint length followed by
+			// the _raw_ TLV bytes.
+			tlvLen := uint64(len(permsTLVBytes.Bytes()))
```
but this is encoding an _array_ of tlv streams. So yes, the tlv layer above will know the length of the whole array but not the lengths of the individual tlv streams. Or am i missing something here?

guggero replied:
Ah I see. And it looks like this is exactly how we do it with the invoices in `lnd`. Sorry, never mind then :sweat_smile:

ellemouton reviewed (commented):
Thanks @guggero 💪 I applied your patch. and also left a comment on your comment about the mac recipe encoding

ellemouton commented:
hmmm what 4 byte id do we want to give the in memory super mac? a fixed one or can we just randomly generate?

guggero reviewed (approved):
Nice, LGTM :100: 

> hmmm what 4 byte id do we want to give the in memory super mac? a fixed one or can we just randomly generate? (made it empty for now)

Yeah, a fixed one that equals `prefix || 0` is what we should use IMO, so basically exactly like you currently implemented it :+1:

ellemouton commented:
This is ready for re-review now that https://github.com/lightninglabs/lightning-node-connect/pull/27 is in 👍
